DQN of dynamic graph version

rical730 opened the pull request:
1. DQN in cartPole
2. add unittest for deepcopy(model) or deepcopy(actor_critic)

rical730 committed:
add dqn based paddle2.0 and add unittest for deepcopy(model or actor_critic)

rical730 committed:
yapf

rical730 commented:
TODO：check paddle version for every fluid example

rical730 committed:
Merge branch 'develop' into dy_dqn

rical730 committed:
format policy gradient

rical730 committed:
replace all fliud api to dygraph api

rical730 committed:
parl>=1.4.2 in readme

rical730 committed:
update paddle version in teamcity

rical730 commented:
> TODO：check paddle version for every fluid example

done in https://github.com/PaddlePaddle/PARL/pull/531

rical730 committed:
update paddle version in windows_unittest

TomorrowIsAnOtherDay committed:
Merge branch 'develop' into dy_dqn

rical730 committed:
Merge branch 'develop' into dy_dqn

TomorrowIsAnOtherDay commented on benchmark/paddle/DQN/replay_memory.py:
```diff
@@ -0,0 +1,46 @@
+#   Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
```
Can we reuse the rpm in `parl.utils`?
https://github.com/PaddlePaddle/PARL/blob/develop/parl/utils/replay_memory.py

rical730 replied:
deleted

TomorrowIsAnOtherDay commented on benchmark/paddle/DQN/README.md:
```diff
@@ -0,0 +1,27 @@
+## Reproduce DQN with PARL
+Based on PARL, we provide a simple demonstration of DQN.
```
we provide a simple demonstration of DQN.
----------->
we provide a simple example of the DQN algorithm.

rical730 replied:
done

TomorrowIsAnOtherDay commented on benchmark/paddle/DQN/README.md:
```diff
@@ -0,0 +1,27 @@
++ [paddlepaddle>=2.0.0](https://github.com/PaddlePaddle/Paddle)
++ [parl>=1.4.2](https://github.com/PaddlePaddle/PARL)
++ gym
++ tqdm
```
Do we need tqdm?

rical730 replied:
removed

TomorrowIsAnOtherDay commented on benchmark/paddle/DQN/cartpole_agent.py:
```diff
@@ -0,0 +1,101 @@
+        reward = np.expand_dims(reward, axis=-1)
+        terminal = np.expand_dims(terminal, axis=-1)
+
+        obs = paddle.to_tensor(obs.astype(np.float32))
```
Why do we need `xxx.as_type(xxx)` here?
Is is possible to remove them?

rical730 replied:
replaced by `, dtype=xxxx`

TomorrowIsAnOtherDay commented on benchmark/paddle/DQN/cartpole_agent.py:
```diff
@@ -0,0 +1,101 @@
+        next_obs = paddle.to_tensor(next_obs.astype(np.float32))
+        terminal = paddle.to_tensor(terminal.astype(np.float32))
+        loss = self.alg.learn(obs, act, reward, next_obs, terminal)
+        return loss.numpy()[0]
```
Please cast the paddle tensor to np.array in the `algorithm` scope rather than the `agent` scope.

rical730 replied:
replaced by `, dtype=xxxx`

TomorrowIsAnOtherDay commented on benchmark/paddle/DQN/cartpole_agent.py:
```diff
@@ -0,0 +1,101 @@
+        self.e_greed_decrement = e_greed_decrement
+
+    def sample(self, obs):
+        """Sample an action when given an observation
```
Sample an action `for exploration` when given an observation

rical730 replied:
done

TomorrowIsAnOtherDay commented on benchmark/paddle/DQN/cartpole_agent.py:
```diff
@@ -0,0 +1,101 @@
+        self.e_greed = max(0.01, self.e_greed - self.e_greed_decrement)
+        return act
+
+    def predict(self, obs):
```
Predict an action `for evaluation` when given an observation

rical730 replied:
done

TomorrowIsAnOtherDay commented on benchmark/paddle/QuickStart/README.md:
```diff
@@ -4,7 +4,7 @@ Train an agent with PARL to solve the CartPole problem, a classical benchmark in
 ### Dependencies:
 
-+ [parl](https://github.com/PaddlePaddle/PARL)
++ [parl>=1.4](https://github.com/PaddlePaddle/PARL)
```
1.4.2

rical730 replied:
policy gradient can be import in parl 1.4

TomorrowIsAnOtherDay commented on parl/algorithms/paddle/dqn.py:
```diff
@@ -0,0 +1,73 @@
+
+        Args:
+            model (parl.Model): model defining forward network of Q function.
+            gamma (float): discounted factor for reward computation.
```
discounted factor for `accumulative` reward computation

rical730 replied:
done

TomorrowIsAnOtherDay commented on parl/algorithms/paddle/dqn.py:
```diff
@@ -0,0 +1,73 @@
+        """ DQN algorithm
+
+        Args:
+            model (parl.Model): model defining forward network of Q function.
```
forward neural network representing the Q function.

rical730 replied:
done

TomorrowIsAnOtherDay commented on parl/algorithms/paddle/dqn.py:
```diff
@@ -0,0 +1,73 @@
+            learning_rate=lr, parameters=self.model.parameters())
+
+    def predict(self, obs):
+        """ use value model self.model to predict the action value
```
use self.model (Q function) to predict the action values

rical730 replied:
done

TomorrowIsAnOtherDay commented on parl/algorithms/paddle/dqn.py:
```diff
@@ -0,0 +1,73 @@
+        return self.model.value(obs)
+
+    def learn(self, obs, action, reward, next_obs, terminal):
+        """ update value model self.model with DQN algorithm
```
update the Q function (self.model) with DQN algorithm

rical730 replied:
done

rical730 committed:
update replay memory and  specify dtype in to_tensor

1. use parl.utils.ReplayMemory (updated, compatible with discrete actions);
2. specify dtype in to_tensor() in DQN/DDPG/PG;
3. update some comments.

rical730 committed:
yapf

rical730 committed:
yapf

TomorrowIsAnOtherDay reviewed (approved):
LGTM.
